src/stepper.c

```diff
@@ -209,39 +212,35 @@ void
 command_queue_step(uint32_t *args)
 {
     struct stepper *s = stepper_oid_lookup(args[0]);
-    struct stepper_move *m = move_alloc();
-    m->interval = args[1];
-    m->count = args[2];
-    if (!m->count)
+    struct stepper_move *new_event = mq_alloc_event();
+    new_event->interval = args[1];
+    new_event->count = args[2];
+    if (!new_event->count)
         shutdown("Invalid count parameter");
-    m->add = args[3];
-    m->next = NULL;
-    m->flags = 0;
+    new_event->add = args[3];
+    new_event->flags = 0;
 
     irq_disable();
     uint8_t flags = s->flags;
     if (!!(flags & SF_LAST_DIR) != !!(flags & SF_NEXT_DIR)) {
         flags ^= SF_LAST_DIR;
-        m->flags |= MF_DIR;
+        new_event->flags |= MF_DIR;
     }
     flags &= ~SF_NO_NEXT_CHECK;
-    if (m->count == 1 && (m->flags || flags & SF_LAST_RESET))
+    if (new_event->count == 1 && (new_event->flags || flags & SF_LAST_RESET))
         // count=1 moves after a reset or dir change can have small intervals
         flags |= SF_NO_NEXT_CHECK;
     flags &= ~SF_LAST_RESET;
-    if (s->count) {
-        s->flags = flags;
-        if (s->first)
-            *s->plast = m;
-        else
-            s->first = m;
-        s->plast = &m->next;
-    } else if (flags & SF_NEED_RESET) {
-        move_free(m);
-    } else {
-        s->flags = flags;
-        s->first = m;
-        stepper_load_next(s, s->next_step_time + m->interval);
+    if (flags & SF_NEED_RESET) {
+        mq_free_event(new_event);
+        irq_enable();
+        return;
+    }
+
+    s->flags = flags;
+    mq_event_insert(&s->queue, &new_event->event);
+    if (!s->count) {
+        stepper_load_next(s, s->next_step_time + new_event->interval);
```

Comment:
I am unsure if this is actually correct in every case, because I don't know when `s->count` is > 0.
The current code inserts the event in every case, but before it inserted it only partially (s->first = m) when count = 0.
Is this an indication that no further events exist, and discards of the other items?

Reply:
Okay - I'll take a closer look at that code (I didn't look in detail).

At a high-level, the check for `s->count > 0` detemines if the stepper is active at all (anything scheduled or currently running).  It differs from a check for `s->first != NULL` in the case where there are no more queued commands pending, but the code is still executing the steps from the last command.